nengo/builder/optimizer.py

```diff
@@ -0,0 +1,911 @@
+        if op not in self.info:
+            try:
+                first_view = next(s for s in op.all_signals if s.is_view)
+                self.info[op] = self._OpDetails(first_view=first_view,
+                                                v_offset=first_view.offset,
+                                                v_size=first_view.nbytes,
+                                                v_base=first_view.base)
+            except StopIteration:
+                self.info[op] = self._OpDetails(
+                    first_view=None, v_offset=0, v_size=0, v_base=None)
+        return self.info[op]
+
+    def __iter__(self):
+        return iter(self.info)
+
+    def __len__(self):
+        return len(self.info)
+
+    def __repr__(self):
+        return repr(self.info)
+
+    def __str__(self):
+        return str(self.info)
+
+    def clear(self):
+        self.info.clear()
+
+
+class OpsToMerge(object):
+    """Analyze and store extra information about a list of ops to be merged."""
+
+    def __init__(self, initial_op, merged, dependents):
+        self.merged = merged
+        self.dependents = dependents
+        self.ops = [initial_op]
+        self.optype = type(initial_op)
+        self.opinfo = OpInfo()
+
+        self.all_signals = set(initial_op.all_signals)
+        self.all_dependents = set(self.dependents[initial_op])
+
+    @property
+    def last_op(self):
+        return self.ops[-1]
+
+    def add(self, op):
+        self.ops.append(op)
+        self.all_signals.update(op.all_signals)
+        self.all_dependents.update(self.dependents[op])
+
+    @staticmethod
+    def check_signals(op1, op2):
+        for s1, s2 in zip(op1.all_signals, op2.all_signals):
+            # If one op's signal is a view, the other must be as well
+            if s1.is_view is not s2.is_view:
+                return False
+
+            if s1.is_view:
+                # Views must be on the same base
+                if s1.base is not s2.base:
+                    return False
+                # Views must have the same dtype
+                elif s1.dtype is not s2.dtype:
+                    return False
+                # Views must have the same strides
+                elif s1.strides != s2.strides:
+                    return False
+
+        return True
+
+    def not_sequential(self, op):
+        lastop = self.opinfo[self.ops[-1]]
+        return (lastop.v_offset + lastop.v_size < self.opinfo[op].v_offset)
+
+
+class OpMerger(object):
+    mergers = {}
+
+    @classmethod
+    def is_mergeable(cls, op, tomerge):
+        merger = cls.mergers[tomerge.optype]
+
+        independent_of_ops_tomerge = (
+            op not in tomerge.all_dependents and
+            len(tomerge.dependents[op].intersection(tomerge.ops)) == 0)
+        independent_of_prior_merges = (
+            op not in tomerge.merged and
+            all(op not in tomerge.dependents[o] and
+                o not in tomerge.dependents[op] for o in tomerge.merged))
```

Comment:
I think that is only a constant factor, while the main problem at the moment is the algorithmic complexity. But I'll probably try it at some point.